src/mds/MDSDaemon.cc

```diff
@@ -1245,24 +1245,40 @@ bool MDSDaemon::ms_handle_reset(Connection *con)
   if (con->get_peer_type() != CEPH_ENTITY_TYPE_CLIENT)
     return false;
 
-  Mutex::Locker l(mds_lock);
-  if (stopping) {
-    return false;
+  // ms_handle_reset may be called within mds_lock 
+  // via 'MDSDaemon::handle_signal' -> 'MDSDaemon::suicide' 
+  // -> 'MDSRankDispatcher::shutdown' -> 'AsyncMessenger::shutdown' 
+  // -> 'AsyncMessenger::mark_down_all' -> ... -> 'AsyncConnection::stop' 
+  // -> dispatch 'C_handle_reset' event -> 'Messenger::ms_deliver_handle_reset'
+  const bool is_locked = mds_lock.is_locked();
```

Comment:
@jcsp No, we can't simply use mds_lock.is_locked_by_me() here because one mds thread will hold mds_lock in MDSDaemon::handle_signal(), later it will dispatch an async event 'C_handle_reset' in AsyncConnection::stop() to another msgr thread which will call MDSDaemon::ms_handle_reset().